File: ios/RNFirebase/perf/RNFirebasePerformance.m
#import "RNFirebasePerformance.h"

#if __has_include(<FirebasePerformance/FIRPerformance.h>)
#import <FirebasePerformance/FIRPerformance.h>
#import <FirebasePerformance/FIRHTTPMetric.h>
@implementation RNFirebasePerformance
RCT_EXPORT_MODULE();

- (id)init {
    self = [super init];
    if (self != nil) {
        _traces = [[NSMutableDictionary alloc] init];
        _httpMetrics = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (FIRHTTPMethod) mapStringToMethod:(NSString *) value {
    if ([value compare:@"get" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodGET;
    if ([value compare:@"put" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodPUT;
    if ([value compare:@"post" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodPUT;
    if ([value compare:@"delete" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodDELETE;
    if ([value compare:@"head" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodHEAD;
    if ([value compare:@"patch" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodPATCH;
    if ([value compare:@"options" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodOPTIONS;
    if ([value compare:@"trace" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodTRACE;
    if ([value compare:@"connect" options:NSCaseInsensitiveSearch] == NSOrderedSame) return FIRHTTPMethodCONNECT;
    return FIRHTTPMethodGET;
}


- (FIRTrace *)getOrCreateTrace:(NSString *)identifier {
    if (_traces[identifier]) {
        return _traces[identifier];
    }
    FIRTrace *trace = [[FIRPerformance sharedInstance] traceWithName:identifier];
    _traces[identifier] = trace;
    return trace;
}

- (FIRHTTPMetric *)getOrCreateHttpMetric:(NSString *)url httpMethod:(NSString *) httpMethod {
    NSString *identifier = [NSString stringWithFormat:@"%@%@", url, httpMethod];
    if (_httpMetrics[identifier]) {
        return _httpMetrics[identifier];
    }
    NSURL * toURL = [NSURL URLWithString:url];
    FIRHTTPMethod method = [self mapStringToMethod:httpMethod];
    FIRHTTPMetric *httpMetric = [[FIRHTTPMetric alloc] initWithURL:toURL HTTPMethod:method];
    _httpMetrics[identifier] = httpMetric;
    return httpMetric;
}

RCT_EXPORT_METHOD(setPerformanceCollectionEnabled:
    (BOOL *) enabled) {
    [FIRPerformance sharedInstance].dataCollectionEnabled = (BOOL) enabled;
}

/**
 * Trace
 */

RCT_EXPORT_METHOD(startTrace:
                  (NSString *) identifier
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateTrace:identifier] start];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(stopTrace:
                  (NSString *) identifier
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateTrace:identifier] stop];
    [_traces removeObjectForKey:identifier];

    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(getTraceAttribute:
                  (NSString *) identifier
                  attribute:(NSString *) attribute
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *value = [[self getOrCreateTrace:identifier] valueForAttribute:attribute];
    resolve(value ? value : [NSNull null]);
}

RCT_EXPORT_METHOD(getTraceAttributes:
                  (NSString *) identifier
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    resolve([[self getOrCreateTrace:identifier] attributes]);
}

RCT_EXPORT_METHOD(getTraceLongMetric:
                  (NSString *) identifier
                  metricName:(NSString *) metricName
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    int64_t value = [[self getOrCreateTrace:identifier] valueForIntMetric:metricName];
    resolve(@(value));
}

RCT_EXPORT_METHOD(incrementTraceMetric:
                  (NSString *) identifier
                  metricName:(NSString *) metricName
                  incrementBy:(nonnull NSNumber *) incrementBy
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    int64_t byInt = [incrementBy intValue];
    [[self getOrCreateTrace:identifier] incrementMetric:metricName byInt:byInt];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(putTraceAttribute:
                  (NSString *) identifier
                  attribute:(NSString *) attribute
                  value:(NSString *) value
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    FIRTrace * trace = [self getOrCreateTrace:identifier];
    [trace setValue:value forAttribute:attribute];
    
    if (trace.attributes[attribute] != nil) {
        resolve(@(YES));
    } else {
        resolve(@(NO));
    }
}

RCT_EXPORT_METHOD(putTraceMetric:
                  (NSString *) identifier
                  attribute:(NSString *) attribute
                  value:(nonnull NSNumber *) value
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    int64_t byInt = [value intValue];
    [[self getOrCreateTrace:identifier] setIntValue:byInt forMetric:attribute];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(removeTraceAttribute:
                  (NSString *) identifier
                  attribute:(NSString *) attribute
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateTrace:identifier] removeAttribute:attribute];
    resolve([NSNull null]);
}

/**
 * HTTP Metric
 */

RCT_EXPORT_METHOD(startHttpMetric:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateHttpMetric:url httpMethod:httpMethod] start];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(stopHttpMetric:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateHttpMetric:url httpMethod:httpMethod] stop];
    [_httpMetrics removeObjectForKey:[NSString stringWithFormat:@"%@%@", url, httpMethod]];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(getHttpMetricAttribute:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  attribute:(NSString *) attribute
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *value = [[self getOrCreateHttpMetric:url httpMethod:httpMethod] valueForAttribute:attribute];
    resolve(value ? value : [NSNull null]);
}

RCT_EXPORT_METHOD(getHttpMetricAttributes:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    resolve([[self getOrCreateHttpMetric:url httpMethod:httpMethod] attributes]);
}

RCT_EXPORT_METHOD(putHttpMetricAttribute:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  attribute:(NSString *) attribute
                  value:(NSString *) value
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    FIRHTTPMetric * httpMetric = [self getOrCreateHttpMetric:url httpMethod:httpMethod];
    [httpMetric setValue:value forAttribute:attribute];
    
    if (httpMetric.attributes[attribute] != nil) {
        resolve(@(YES));
    } else {
        resolve(@(NO));
    }
}

RCT_EXPORT_METHOD(removeHttpMetricAttribute:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  attribute:(NSString *) attribute
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateHttpMetric:url httpMethod:httpMethod] removeAttribute:attribute];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(setHttpMetricResponseCode:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  code:(nonnull NSNumber *) code
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateHttpMetric:url httpMethod:httpMethod] setResponseCode:[code integerValue]];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(setHttpMetricRequestPayloadSize:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  bytes:(nonnull NSNumber *) bytes
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateHttpMetric:url httpMethod:httpMethod] setRequestPayloadSize:[bytes longLongValue]];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(setHttpMetricResponseContentType:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  type:(NSString *) type
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateHttpMetric:url httpMethod:httpMethod] setResponseContentType:type];
    resolve([NSNull null]);
}

RCT_EXPORT_METHOD(setHttpMetricResponsePayloadSize:
                  (NSString *) url
                  httpMethod:(NSString *) httpMethod
                  bytes:(nonnull NSNumber *) bytes
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [[self getOrCreateHttpMetric:url httpMethod:httpMethod] setResponsePayloadSize:[bytes longLongValue]];
    resolve([NSNull null]);
}

+ (BOOL)requiresMainQueueSetup
{
    return YES;
}

@end

#else
@implementation RNFirebasePerformance
@end
#endif

File: android/src/main/java/io/invertase/firebase/perf/RNFirebasePerformance.java
package io.invertase.firebase.perf;


import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.google.firebase.perf.FirebasePerformance;
import com.google.firebase.perf.metrics.HttpMetric;
import com.google.firebase.perf.metrics.Trace;

import java.util.HashMap;
import java.util.Map;

public class RNFirebasePerformance extends ReactContextBaseJavaModule {

  private static final String TAG = "RNFirebasePerformance";
  private HashMap<String, Trace> traces = new HashMap<>();
  private HashMap<String, HttpMetric> httpMetrics = new HashMap<>();

  public RNFirebasePerformance(ReactApplicationContext reactContext) {
    super(reactContext);
    Log.d(TAG, "New instance");
  }

  /**
   * @return
   */
  @Override
  public String getName() {
    return TAG;
  }

  @ReactMethod
  public void setPerformanceCollectionEnabled(Boolean enabled, Promise promise) {
    FirebasePerformance.getInstance().setPerformanceCollectionEnabled(enabled);
    promise.resolve(null);
  }

  /**
   * Trace
   */

  @ReactMethod
  public void getTraceAttribute(String identifier, String attribute, Promise promise) {
    promise.resolve(getOrCreateTrace(identifier).getAttribute(attribute));
  }

  @ReactMethod
  public void getTraceAttributes(String identifier, Promise promise) {
    Map<String, String> attributes = getOrCreateTrace(identifier).getAttributes();
    WritableMap map = Arguments.createMap();

    for (Map.Entry<String, String> entry : attributes.entrySet()) {
      map.putString(entry.getKey(), entry.getValue());
    }

    promise.resolve(map);
  }

  @ReactMethod
  public void getTraceLongMetric(String identifier, String metricName, Promise promise) {
    Integer value = Long.valueOf(getOrCreateTrace(identifier).getLongMetric(metricName)).intValue();
    promise.resolve(value);
  }

  @ReactMethod
  public void incrementTraceMetric(
    String identifier,
    String metricName,
    Integer incrementBy,
    Promise promise
  ) {
    getOrCreateTrace(identifier).incrementMetric(metricName, incrementBy.longValue());
    promise.resolve(null);
  }

  @ReactMethod
  public void putTraceAttribute(
    String identifier,
    String attribute,
    String value,
    Promise promise
  ) {
    getOrCreateTrace(identifier).putAttribute(attribute, value);
    // Docs say it returns a bool, actually void so we internally check attributes
    Map<String, String> attributes = getOrCreateTrace(identifier).getAttributes();
    if (attributes.containsKey(attribute)) {
      promise.resolve(true);
    } else {
      promise.resolve(false);
    }
  }

  @ReactMethod
  public void putTraceMetric(String identifier, String metricName, Integer value, Promise promise) {
    getOrCreateTrace(identifier).putMetric(metricName, value.longValue());
    promise.resolve(null);
  }

  @ReactMethod
  public void removeTraceAttribute(String identifier, String attribute, Promise promise) {
    getOrCreateTrace(identifier).removeAttribute(attribute);
    promise.resolve(null);
  }

  @ReactMethod
  public void startTrace(String identifier, Promise promise) {
    getOrCreateTrace(identifier).start();
    promise.resolve(null);
  }

  @ReactMethod
  public void stopTrace(String identifier, Promise promise) {
    getOrCreateTrace(identifier).stop();
    traces.remove(identifier);
    promise.resolve(null);
  }

  /**
   * Http Metric
   */

  @ReactMethod
  public void getHttpMetricAttribute(
    String url,
    String httpMethod,
    String attribute,
    Promise promise
  ) {
    promise.resolve(getOrCreateHttpMetric(url, httpMethod).getAttribute(attribute));
  }

  @ReactMethod
  public void getHttpMetricAttributes(String url, String httpMethod, Promise promise) {
    Map<String, String> attributes = getOrCreateHttpMetric(url, httpMethod).getAttributes();
    WritableMap map = Arguments.createMap();

    for (Map.Entry<String, String> entry : attributes.entrySet()) {
      map.putString(entry.getKey(), entry.getValue());
    }

    promise.resolve(map);
  }

  @ReactMethod
  public void putHttpMetricAttribute(
    String url,
    String httpMethod,
    String attribute,
    String value,
    Promise promise
  ) {
    getOrCreateHttpMetric(url, httpMethod).putAttribute(attribute, value);
    // Docs say it returns a bool, actually void so we internally check attributes
    Map<String, String> attributes = getOrCreateHttpMetric(url, httpMethod).getAttributes();
    if (attributes.containsKey(attribute)) {
      promise.resolve(true);
    } else {
      promise.resolve(false);
    }
  }

  @ReactMethod
  public void removeHttpMetricAttribute(
    String url,
    String httpMethod,
    String attribute,
    Promise promise
  ) {
    getOrCreateHttpMetric(url, httpMethod).removeAttribute(attribute);
    promise.resolve(null);
  }

  @ReactMethod
  public void setHttpMetricResponseCode(
    String url,
    String httpMethod,
    Integer code,
    Promise promise
  ) {
    getOrCreateHttpMetric(url, httpMethod).setHttpResponseCode(code);
    promise.resolve(null);
  }

  @ReactMethod
  public void setHttpMetricRequestPayloadSize(
    String url,
    String httpMethod,
    Integer bytes,
    Promise promise
  ) {
    getOrCreateHttpMetric(url, httpMethod).setRequestPayloadSize(bytes.longValue());
    promise.resolve(null);
  }

  @ReactMethod
  public void setHttpMetricResponseContentType(
    String url,
    String httpMethod,
    String type,
    Promise promise
  ) {
    getOrCreateHttpMetric(url, httpMethod).setResponseContentType(type);
    promise.resolve(null);
  }

  @ReactMethod
  public void setHttpMetricResponsePayloadSize(
    String url,
    String httpMethod,
    Integer bytes,
    Promise promise
  ) {
    getOrCreateHttpMetric(url, httpMethod).setResponsePayloadSize(bytes.longValue());
    promise.resolve(null);
  }

  @ReactMethod
  public void startHttpMetric(String url, String httpMethod, Promise promise) {
    getOrCreateHttpMetric(url, httpMethod).start();
    promise.resolve(null);
  }

  @ReactMethod
  public void stopHttpMetric(String url, String httpMethod, Promise promise) {
    getOrCreateHttpMetric(url, httpMethod).stop();
    httpMetrics.remove(url + httpMethod);
    promise.resolve(null);
  }

  /**
   * Private
   */

  private Trace getOrCreateTrace(String identifier) {
    if (traces.containsKey(identifier)) {
      return traces.get(identifier);
    }
    Trace trace = FirebasePerformance.getInstance().newTrace(identifier);
    traces.put(identifier, trace);
    return trace;
  }

  private HttpMetric getOrCreateHttpMetric(String url, String httpMethod) {
    String identifier = url + httpMethod;
    if (httpMetrics.containsKey(identifier)) {
      return httpMetrics.get(identifier);
    }
    HttpMetric httpMetric = FirebasePerformance.getInstance().newHttpMetric(
      url,
      this.mapStringToMethod(httpMethod)
    );
    httpMetrics.put(identifier, httpMetric);
    return httpMetric;
  }

  private String mapStringToMethod(String value) {
    switch (value) {
      case "CONNECT":
        return FirebasePerformance.HttpMethod.CONNECT;
      case "DELETE":
        return FirebasePerformance.HttpMethod.DELETE;
      case "GET":
        return FirebasePerformance.HttpMethod.GET;
      case "HEAD":
        return FirebasePerformance.HttpMethod.HEAD;
      case "OPTIONS":
        return FirebasePerformance.HttpMethod.OPTIONS;
      case "PATCH":
        return FirebasePerformance.HttpMethod.PATCH;
      case "POST":
        return FirebasePerformance.HttpMethod.POST;
      case "PUT":
        return FirebasePerformance.HttpMethod.PUT;
      case "TRACE":
        return FirebasePerformance.HttpMethod.TRACE;
    }

    return "";
  }
}
